### PR TITLE
Functionality to allow connector to assume an AWS IAM role when invok…

### PR DESCRIPTION
**Pull request by seananthonywilliams**
…ing a lambda

### CODE REVIEW

**Comment by dylanmei** [comment]
Related to #4. @SgtPepperLHCB please have a look.

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (dylanmei) [inline LICENSE]: This should be in its own commit; it's not related to assuming a role.

File: pom.xml [hunk 2]
```diff
@@ -39,23 +43,34 @@
...
+            <version>${mockito-core.version}</version>
         </dependency>
         <dependency>
             <groupId>com.google.guava</groupId>
             <artifactId>guava</artifactId>
-            <version>19.0</version>
+            <version>${google.guava.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
```
Comment (dylanmei) [inline pom.xml]: This dependency shouldn't be necessary.

File: README.md [hunk 3]
```diff
@@ -18,6 +18,13 @@ A sink connector reads from a Kafka topic and sends events to an AWS Lambda func
 A sink connector configuration has two required fields:
  * `aws.lambda.function.arn`: The AWS ARN of the Lambda function to send events to.
  * `topics`: The Kafka topic to be read from.
+ 
+### AWS Assume Role Support options
+ The connector can assume a cross-account role:
```
Comment (dylanmei) [inline README.md]: It doesn't have to be "cross-account" to be useful. Consider:

> The connector can assume an IAM Role. The role must include a policy that allows `lambda:InvokeFunction` and `lambda:InvokeAsync` actions.

File: README.md [hunk 4]
```diff
@@ -18,6 +18,13 @@ A sink connector reads from a Kafka topic and sends events to an AWS Lambda func
...
+ * `lambda.credentials.provider.class=com.nordstrom.kafka.connect.auth.AWSAssumeRoleCredentialsProvider`: REQUIRED Class providing cross-account role assumption.
```
Comment (dylanmei) [inline README.md]: Can omit "cross-account". Consider:

> REQUIRED The credentials provider class.

File: LICENSE (hunk 1 above)
Comment (dylanmei) [reply LICENSE]: Separate issue added here: https://github.com/Nordstrom/kafka-connect-lambda/issues/8

File: src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfig.java [hunk 5]
```diff
@@ -259,31 +293,58 @@ public static ConfigDef config() {
             + RETRIABLE_ERROR_CODES_DEFAULT),
     RETRY_BACKOFF_MILLIS("retry.backoff.millis",
             "The amount of time to wait between retry attempts, by default is "
-                    + RETRY_BACKOFF_MILLIS_DEFAULT);
+                    + RETRY_BACKOFF_MILLIS_DEFAULT),
 
+    // AWS assume role support options
+    CREDENTIALS_PROVIDER_CLASS_CONFIG("lambda.credentials.provider.class", "REQUIRED Class providing cross-account role assumption"),
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfig.java]: Earlier configs are prefixed with `aws.` but these are not.

File: src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java [hunk 6]
```diff
@@ -32,37 +36,41 @@
     private final AWSLambdaAsync lambdaClient;
     private final InvocationFailure failureMode;
 
-    public AwsLambdaUtil(final Configuration config) {
-        Guard.verifyNotNull(config, "config");
+    public AwsLambdaUtil(final Configuration optConfigs, final Map<String, ?> bareAssumeRoleConfigs) {
+        LOGGER.warn("AwsLambdaUtil.ctor:bareAssumeRoleConfigs={}", bareAssumeRoleConfigs);
```
Comment (SgtPepperLHCB) [inline src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java]: Are these log statements really WARN level or DEBUG/TRACE?

Comment (seananthonywilliams) [reply src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java]: Updated log statements to reflect that they should be info vs. warn.

File: LICENSE (hunk 1 above)
Comment (seananthonywilliams) [reply LICENSE]: Removed.

File: pom.xml (hunk 2 above)
Comment (seananthonywilliams) [reply pom.xml]: Removed.  I mainly added it to avoid having to write/generate explicit getters for the externalId, roleArn, and sessionName instance variables.

File: README.md (hunk 4 above)
Comment (seananthonywilliams) [reply README.md]: Updated.

File: README.md (hunk 3 above)
Comment (seananthonywilliams) [reply README.md]: Updated.

File: src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfig.java (hunk 5 above)
Comment (seananthonywilliams) [reply src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfig.java]: Hmm... I had actually opted for "lambda.*" prefixes to be consistent with our SQS Lambda connector IAM assume role support options [here](https://github.com/Nordstrom/kafka-connect-sqs/blob/master/README.md#aws-assume-role-support-options).

Comment (dylanmei) [reply src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfig.java]: Let's chose to be consistent with this project, or that project over there. If you prefer the latter, support both but mark the originals deprecated -- to be removed in some future version.

Comment (seananthonywilliams) [reply src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfig.java]: Ah shoot.  I see what you're saying, as the others all have "aws.lambda.*" in front.

Comment (seananthonywilliams) [reply src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfig.java]: I would err on not removing "aws.*" from the priors and adding it to the newers, to not annoy users who may be using this currently.

Comment (dylanmei) [reply src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkConnectorConfig.java]: :+1:

File: src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkTask.java [hunk 7]
```diff
@@ -58,9 +69,9 @@ public void start(Map<String, String> props) {
   }
 
   @Override
-  public void put(Collection<SinkRecord> records) {
+  public void put(final Collection<SinkRecord> records) {
     if (records == null || records.isEmpty()) {
-      LOGGER.debug("No records to process.  connector=\"{}\" task=\"{}\"",
+      LOGGER.info("No records to process.  connector=\"{}\" task=\"{}\"",
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkTask.java]: `debug` seemed like the appropriate level for this.

File: src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkTask.java [hunk 8]
```diff
@@ -70,15 +81,16 @@ public void put(Collection<SinkRecord> records) {
       this.batchRecords.addAll(records);
       final int batchLength = this.getPayload(this.batchRecords).getBytes().length;
       if (batchLength >= this.configuration.getMaxBatchSizeBytes()) {
-        LOGGER.warn("Batch size reached {} bytes within {} records. connector=\"{}\" task=\"{}\"",
+        LOGGER.info("Batch size reached {} bytes within {} records. connector=\"{}\" task=\"{}\"",
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkTask.java]: `warn` seemed like the appropriate level for this.

File: src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java [hunk 9]
```diff
@@ -32,37 +36,41 @@
     private final AWSLambdaAsync lambdaClient;
     private final InvocationFailure failureMode;
 
-    public AwsLambdaUtil(final Configuration config) {
-        Guard.verifyNotNull(config, "config");
+    public AwsLambdaUtil(final Configuration optConfigs, final Map<String, ?> bareAssumeRoleConfigs) {
+        LOGGER.info("AwsLambdaUtil.ctor:bareAssumeRoleConfigs={}", bareAssumeRoleConfigs);
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java]: This seems like a debugging message. Change to `debug` or simply delete this.

File: src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java [hunk 10]
```diff
@@ -147,6 +155,41 @@ InvocationResponse checkPayloadSizeForInvocationType(final byte[] payload, final
...
+    public AWSCredentialsProvider getCredentialsProvider(Map<String, ?> roleConfigs) {
+        LOGGER.info(".get-credentials-provider:assumeRoleConfigs={}", roleConfigs);
+
+        try {
+            Object providerField = roleConfigs.get("class");
+            String providerClass =  LambdaSinkConnectorConfig.ConfigurationKeys.CREDENTIALS_PROVIDER_CLASS_DEFAULT.getValue();
+            if (providerField != null) {
+                providerClass = providerField.toString();
+            }
+            LOGGER.info(".get-credentials-provider:field={}, class={}", providerField, providerClass);
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java]: This seems like a debugging message. Change to `debug` or simply delete this.

File: src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java [hunk 11]
```diff
@@ -147,6 +155,41 @@ InvocationResponse checkPayloadSizeForInvocationType(final byte[] payload, final
...
+            AWSCredentialsProvider provider = ((Class<? extends AWSCredentialsProvider>)
+                    getClass(providerClass)).newInstance();
+
+            if (provider instanceof Configurable) {
+                ((Configurable) provider).configure(roleConfigs);
+            }
+
+            LOGGER.info(".get-credentials-provider:provider={}", provider);
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java]: This seems like a debugging message. Change to `debug` or simply delete this.

File: src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java [hunk 12]
```diff
@@ -147,6 +155,41 @@ InvocationResponse checkPayloadSizeForInvocationType(final byte[] payload, final
...
+            return provider;
+        } catch (IllegalAccessException | InstantiationException e) {
+            throw new ConnectException("Invalid class for: " + LambdaSinkConnectorConfig.ConfigurationKeys.CREDENTIALS_PROVIDER_CLASS_CONFIG, e);
+        }
+    }
+
+    public Class<?> getClass(String className) {
+        LOGGER.info(".get-class:class={}",className);
```
Comment (dylanmei) [inline src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java]: This seems like a debugging message. Change to `debug` or simply delete this.

**Comment by dylanmei** [comment]
I made a comment regarding verbosity of the logging, much of which appears to be intended for tracing, where chatty logs are most appropriate.

File: src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkTask.java (hunk 7 above)
Comment (seananthonywilliams) [reply src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkTask.java]: Fixed.

File: src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkTask.java (hunk 8 above)
Comment (seananthonywilliams) [reply src/main/java/com/nordstrom/kafka/connect/lambda/LambdaSinkTask.java]: Changed back.

File: src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java (hunk 10 above)
Comment (seananthonywilliams) [reply src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java]: Changed.

File: src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java (hunk 12 above)
Comment (seananthonywilliams) [reply src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java]: Changed.

File: src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java (hunk 11 above)
Comment (seananthonywilliams) [reply src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java]: Changed.

File: src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java (hunk 9 above)
Comment (seananthonywilliams) [reply src/main/java/com/nordstrom/kafka/connect/lambda/AwsLambdaUtil.java]: Changed.
